packages/language-server/src/quick-fix.ts

```diff
@@ -0,0 +1,86 @@
+import {
+  Diagnostic,
+  Range as LSPRange,
+  TextDocument,
+  CodeAction,
+  Command,
+  CodeActionKind,
+  TextDocumentEdit,
+  TextEdit,
+} from "vscode-languageserver-types";
+import { computeQuickFixStableIdInfo } from "@ui5-language-assistant/xml-views-quick-fix";
+import { parse, DocumentCstNode } from "@xml-tools/parser";
+import { buildAst, XMLDocument } from "@xml-tools/ast";
+import { LSPRangeToOffsetRange, offsetRangeToLSPRange } from "./range-utils";
+
+export const QUICK_FIX_STABLE_ID = "nonStableIdQuickFix";
+const QUICK_FIX_STABLE_ID_TITLE = "Generate ID";
+
+export function getQuickFixCodeAction(
+  document: TextDocument,
+  diagnostic: Diagnostic
+): CodeAction | undefined {
+  const documentText = document.getText();
+  // We prefer to parse the document again to avoid cache state handling
+  const { cst, tokenVector } = parse(documentText);
+  const xmlDocAst = buildAst(cst as DocumentCstNode, tokenVector);
+  switch (diagnostic.code) {
+    case 1000: {
+      // non stable id
+      return computeCodeActionForQuickFixStableId({
+        document,
+        xmlDocument: xmlDocAst,
+        nonStableIdDiagnostic: diagnostic,
+      });
+    }
+    default:
+      return undefined;
+  }
+}
+
+function computeCodeActionForQuickFixStableId(opts: {
+  document: TextDocument;
+  xmlDocument: XMLDocument;
+  nonStableIdDiagnostic: Diagnostic;
+}): CodeAction | undefined {
+  const errorOffset = LSPRangeToOffsetRange(
+    opts.nonStableIdDiagnostic.range,
+    opts.document
+  );
+
+  const quickFixIdInfo = computeQuickFixStableIdInfo(
+    opts.xmlDocument,
+    errorOffset
+  );
+  if (quickFixIdInfo === undefined) {
+    return undefined;
+  }
+
+  const title = QUICK_FIX_STABLE_ID_TITLE;
+  return CodeAction.create(
+    title,
+    Command.create(
```

Comment:
This is a "bad" API because of 5 params.
Its not easy to understand what each param is.

- Ensure your const names are "optimal".
- `offsetRangeToLSPRange(quickFixIdInfo.offsetRange, opts.document)` to temp variable with meaningful name?

Reply:
a word with `replace` may fit the extracted variable imho.